examples/nested_list/src/list.rs

```diff
@@ -62,22 +60,19 @@ pub enum Msg {
 pub struct Props {
     pub children: ChildrenRenderer<ListVariant>,
     pub on_hover: Callback<Hovered>,
-    pub weak_link: WeakComponentLink<List>,
 }
 
 pub struct List {
     inactive: bool,
 }
 
-impl Component for List {
+impl BaseComponent for List {
```

Comment:
When I added `BaseComponent`, it is considered to be an internal interface and never intended for end-users.
This means that we can introduce breaking changes to it (like ones that only make sense for function components) without having to consider whether this would have any impact on users. (Given semantic versioning is followed.)

If we recommend users to use BaseComponent (as in the example), we need to make BaseComponent somewhat stable. So if we want to introduce another associated types that is not used by the majority of components or something that is not intended for end-users, we need to introduce a base-`BaseComponent` trait.

If this is the approach that we settle with by associating a type on the component trait, I think we should introduce a `ComponentWithRef` trait that maps onto `BaseComponent` instead of suggesting using `BaseComponent` directly.

Reply:
I agree. `BaseComponent` should *not* be implemented directly. It is exposed so function components can use it (through the macro) and it allows third-party crates to provide different types of components, such as generator components